### PR TITLE
[TRIx-Engine] Add TRIx-Engine subplugin

### PR DESCRIPTION
**Pull request by dongju-chae**
This patch adds TRIx-Engine tensor filter subplugin implementation.
It requires TRIV2 NPU HW and devel packages from the upstream tizen repo.

Signed-off-by: Dongju Chae <dongju.chae@samsung.com>

**Self evaluation:**
1. Build test: [*]Passed [ ]Failed []Skipped
2. Run test: [*]Passed [ ]Failed []Skipped

### CODE REVIEW

**Comment by dongju-chae** [comment]
@dofmind @niley7464 Please check this subplugin to see how nnstreamer and npu-engine are connected.

File: ext/nnstreamer/tensor_filter/tensor_filter_trix_engine.hh [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+/* npu-engine headers */
+#include <npubinfmt.h>
+#include <libnpuhost.h>
+
+/* nnstreamer plugin api headers */
+#include <nnstreamer_plugin_api.h>
+#include <nnstreamer_cppplugin_api_filter.hh>
+#include <nnstreamer_util.h>
+
+#if defined(__TIZEN__)
```
Comment (jaeyun-jung) [inline ext/nnstreamer/tensor_filter/tensor_filter_trix_engine.hh:26]: plz check log functions (is it necessary to define new macro? (e.g. use nnstreamer_log.h)

File: ext/nnstreamer/tensor_filter/tensor_filter_trix_engine.cc [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+
+  input_layout_ = &(prop->input_layout[0]);
+  output_layout_ = &(prop->output_layout[0]);
+}
+
+/**
+ * @brief Convert data layout (from NNStreamer to TRIx-Engine)
+ */
+static data_layout
+convert_data_layout (const tensor_layout &layout) {
```
Comment (jaeyun-jung) [inline ext/nnstreamer/tensor_filter/tensor_filter_trix_engine.cc:144]: convert_data_layout() and convert_data_type()
static or private method in TensorFilterTRIxEngine cls

File: ext/nnstreamer/tensor_filter/tensor_filter_trix_engine.hh (hunk 1 above)
Comment (dongju-chae) [reply ext/nnstreamer/tensor_filter/tensor_filter_trix_engine.hh:26]: Removed, thanks.

File: ext/nnstreamer/tensor_filter/tensor_filter_trix_engine.cc (hunk 2 above)
Comment (dongju-chae) [reply ext/nnstreamer/tensor_filter/tensor_filter_trix_engine.cc:144]: Changed to static method, thanks.

File: ext/nnstreamer/tensor_filter/tensor_filter_trix_engine.cc [hunk 3]
```diff
@@ -0,0 +1,353 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+/**
+ * Copyright (C) 2020 Samsung Electronics
+ * Copyright (C) 2020 Dongju Chae <dongju.chae@samsung.com>
+ */
+/**
+ * @file tensor_filter_trix_engine.cc
+ * @date 20 Jan 2020
```
Comment (gichan-jang) [inline ext/nnstreamer/tensor_filter/tensor_filter_trix_engine.cc:8]: Need to update date.

Comment (dongju-chae) [reply ext/nnstreamer/tensor_filter/tensor_filter_trix_engine.cc:8]: The date should be the original date. The files were just moved from npu-engine repo.

File: ext/nnstreamer/tensor_filter/tensor_filter_trix_engine.cc [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+  status = createNPU_request (dev_, model_id_, &req_id);
+  if (status != 0) {
+    nns_loge ("Unable to create NPU request with model id (%u): %d", model_id_, status);
+    return;
+  }
+
+  tensors_data_info in_info;
+  tensors_data_info out_info;
+  /* set data info using metadata */
+  set_data_info (&in_info, &out_info);
```
Comment (jaeyun-jung) [inline ext/nnstreamer/tensor_filter/tensor_filter_trix_engine.cc:252]: how about calling set_data_info() in configure_instance()?
(define member val in_info_ / out_info_ and remove tensor_layout) 
IMO, converting tensors info can be done in init state.
```
private:
  tensors_data_info in_info_;
  tensors_data_info out_info_;

//below layout val is unnecessary. 
  const tensor_layout *input_layout_;  /**< The data layout of input tensors */ 
  const tensor_layout *output_layout_; /**< The data layout of output tensors */

void set_data_info (const GstTensorFilterProperties *prop)

```

File: ext/nnstreamer/tensor_filter/tensor_filter_trix_engine.cc (hunk 3 above)
Comment (myungjoo) [reply ext/nnstreamer/tensor_filter/tensor_filter_trix_engine.cc:8]: Code in nnstreamer.git is LGPL 2.1 only. Update it.

File: ext/nnstreamer/tensor_filter/tensor_filter_trix_engine.hh [hunk 5]
```diff
@@ -0,0 +1,78 @@
+/* SPDX-License-Identifier: Apache-2.0 */
```
Comment (myungjoo) [inline ext/nnstreamer/tensor_filter/tensor_filter_trix_engine.hh:1]: LGPL 2.1 only

File: packaging/nnstreamer.spec [hunk 6]
```diff
@@ -35,6 +35,7 @@
 %define		lua_support 1
 %define		tvm_support 1
 %define		snpe_support 1
+%define		trix_engine_support 1
```
Comment (myungjoo) [inline packaging/nnstreamer.spec:38]: Set it 0 if it's 
- DA build (LINE 110)
- Tizen 6.5 or lower (around LINE 43-70)

File: ext/nnstreamer/tensor_filter/tensor_filter_trix_engine.hh [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+  void invoke (const GstTensorMemory *input, GstTensorMemory *output);
+  void getFrameworkInfo (GstTensorFilterFrameworkInfo &info);
+  int getModelInfo (model_info_ops ops, GstTensorsInfo &in_info, GstTensorsInfo &out_info);
+  int eventHandler (event_ops ops, GstTensorFilterFrameworkEventData &data);
+
+  /* static methods */
+  static void init_filter_trix_engine ();
+  static void fini_filter_trix_engine ();
+
+  static data_layout convert_data_layout (const tensor_layout &layout);
```
[the remark this answers is not in the document]
Comment (dongju-chae) [reply ext/nnstreamer/tensor_filter/tensor_filter_trix_engine.hh:50]: Updated, thanks.

File: packaging/nnstreamer.spec (hunk 6 above)
Comment (dongju-chae) [reply packaging/nnstreamer.spec:38]: Updated, thanks.

File: ext/nnstreamer/tensor_filter/tensor_filter_trix_engine.cc (hunk 3 above)
Comment (dongju-chae) [reply ext/nnstreamer/tensor_filter/tensor_filter_trix_engine.cc:8]: Updated, thanks.

File: ext/nnstreamer/tensor_filter/tensor_filter_trix_engine.cc (hunk 4 above)
Comment (dongju-chae) [reply ext/nnstreamer/tensor_filter/tensor_filter_trix_engine.cc]: Looks better, thanks.

File: ext/nnstreamer/tensor_filter/tensor_filter_trix_engine.hh (hunk 5 above)
Comment (dongju-chae) [reply ext/nnstreamer/tensor_filter/tensor_filter_trix_engine.hh]: Updated, thanks.

File: ext/nnstreamer/tensor_filter/tensor_filter_trix_engine.hh [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+ * @file    tensor_filter_subplugin_trix_engine.hh
+ * @date    20 Jan 2020
+ * @brief   NNStreamer tensor-filter subplugin trix_engine header
+ * @see     https://github.com/nnstreamer/nnstreamer
+ * @author  Dongju Chae <dongju.chae@samsung.com>
+ * @bug     No known bugs
+ */
+
+#ifndef __TENSOR_FILTER_SUBPLUGIN_TRIxEngine_H__
+
```
Comment (jaeyun-jung) [inline ext/nnstreamer/tensor_filter/tensor_filter_trix_engine.hh:16]: plz add 
`#define __TENSOR_FILTER_SUBPLUGIN_TRIxEngine_H__`

and update line 78
`#endif /* __TENSOR_FILTER_SUBPLUGIN_TRIxEngine_H__ */`

Comment (dongju-chae) [reply ext/nnstreamer/tensor_filter/tensor_filter_trix_engine.hh:17]: My bad :) Thanks.

**Review by jaeyun-jung — APPROVED** [review]

**Review by gichan-jang — APPROVED** [review]

**Review by myungjoo — APPROVED** [review]

**Pull request merged**